### PR TITLE
Fix for NIFI-1838,  NIFI-1152, NIFI-2117

### PR DESCRIPTION
**Pull request by PuspenduBanerjee**
Fix for NIFI-1838,  NIFI-1152, NIFI-2117

### CODE REVIEW

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java [hunk 1]
```diff
@@ -587,13 +589,15 @@ public void transfer(final FlowFile flowFile, final Relationship relationship) {
...
         validateState(flowFile);
+        validateDestinationRelation(relationship);
         List<MockFlowFile> list = transferMap.get(relationship);
         if (list == null) {
             list = new ArrayList<>();
             transferMap.put(relationship, list);
         }
 
         beingProcessed.remove(flowFile.getId());
+       // if(sharedState.)
```
Comment (pvillard31) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: Could you remove this comment?

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java [hunk 2]
```diff
@@ -793,13 +798,21 @@ private void validateState(final FlowFile flowFile) {
...
         for (final List<MockFlowFile> flowFiles : transferMap.values()) {
             if (flowFiles.contains(flowFile)) {
                 throw new IllegalStateException(flowFile + " has already been transferred");
             }
         }
     }
 
+    private void validateDestinationRelation(final Relationship relationship) {
+        if (!relationships.contains(relationship)) {
+            rollback();
```
Comment (pvillard31) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: Not sure to see why a rollback is needed here.

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java [hunk 3]
```diff
@@ -16,6 +16,9 @@
  */
 package org.apache.nifi.util;
 
+import java.util.Collections;
+import java.util.Set;
+
```
Comment (pvillard31) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: I believe your IDE is automatically changing the import order when you save your modifications. Could you revert the changes?

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java [hunk 4]
```diff
@@ -36,7 +36,7 @@ public void testPenalizeFlowFileFromProcessor() {
 
     protected static class PoorlyBehavedProcessor extends AbstractProcessor {
 
-        private static final Relationship REL_FAILURE = new Relationship.Builder()
+        protected static final Relationship REL_FAILURE = new Relationship.Builder()
```
Comment (pvillard31) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: scope change is not needed, right?

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java [hunk 5]
```diff
@@ -55,4 +55,35 @@ public void onTrigger(final ProcessContext ctx, final ProcessSession session) th
...
+
+            static final Relationship REL_NON_EXISTENT = new Relationship.Builder()
+                    .name("NonExistent Relation")
+                    .build();
+
+            public Set<Relationship> relationships = Collections.singleton(REL_FAILURE);
+
+            @Override
+            public Set<Relationship> getRelationships() {
+                return relationships;
```
Comment (pvillard31) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: I think you just have to return an empty set of relationships instead of referencing the one of the other class. No?

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java (hunk 1 above)
Comment (PuspenduBanerjee) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: done

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java (hunk 2 above)
Comment (PuspenduBanerjee) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: So, if a relationship does not exist what should we do?

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java (hunk 4 above)
Comment (PuspenduBanerjee) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: Directly using it in Test NonExistentRelationTestProcessor public Set<Relationship> relationships = Collections.singleton(REL_FAILURE);

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java (hunk 5 above)
Comment (PuspenduBanerjee) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: The way groovy test cases are written , they will fail if we return Empty Relationship. please see NIFI-1838

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java (hunk 2 above)
Comment (pvillard31) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: This class is limited to testing purpose, so throwing an exception would make fail the associated test. I think this is what we want and in this case rollback is not needed. In a real context we should not be in such situation where the relationship does not exist.

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java (hunk 5 above)
Comment (pvillard31) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: It looks to me this is unrelated. This is a test class absolutely independent from scripting processors, you don't need to change the scope of the relationship in ``PoorlyBehavedProcessor``and you can return an empty set of relationships to execute your unit test.

File: nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java [hunk 6]
```diff
@@ -92,11 +92,10 @@
...
                 }
             }
-        } else {
-            // Return defaults for now
-            relationships.add(REL_SUCCESS);
-            relationships.add(REL_FAILURE);
         }
+        // Add defaults
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_FAILURE);
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: @mattyb149 do you mind having a look? I don't think this is what we want here, I'd rather say that if we need to change something it would be on tests side.

Comment (PuspenduBanerjee) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: @pvillard31 Please check the use of session.transfer(flowFile, InvokeScriptedProcessor.REL_SUCCESS) in test_reader.groovy#onTrigger. All groovy test cases uses REL_SUCCESS or REL_FAILURE . It is not a problem in pre-NIFI-1152 and that's why I had to raise NIFI-1838.

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java (hunk 5 above)
Comment (PuspenduBanerjee) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: Good finding. Can't really remember why did I add that. removed now and checking in.

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java (hunk 2 above)
Comment (PuspenduBanerjee) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: I think, though it's for testing purpose , we should emulate as real as we can and rollback is the best possible way to revert back to a stable state. And, in a real context we may face such issue in case of orchestration, where an attribute in a flowfile may finally dictate where it should be transferred.

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java (hunk 4 above)
Comment (PuspenduBanerjee) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: No not needed anymore. :)

**Comment by PuspenduBanerjee** [comment]
@pvillard31 Thanks for your effort to review this. 
I have incorporated your comments as much as I could.

**Comment by PuspenduBanerjee** [comment]
@pvillard31 @mattyb149  Do we still need more discussion on this or will you be able to check it in?

**Comment by pvillard31** [comment]
thanks @PuspenduBanerjee, I'm quite busy this WE, will have another look asap

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java (hunk 4 above)
Comment (pvillard31) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: Is it still needed?

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java (hunk 2 above)
Comment (pvillard31) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: Fair enough.

File: nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java (hunk 6 above)
Comment (mattyb149) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: Yeah if this is failing tests, I'd say take a look at the tests. We don't want to add default relationships, especially when there's a legit instance of a Processor, as it is the Processor's responsibility to define all relationships

Comment (PuspenduBanerjee) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: As per documentation of [ExecuteScript](https://github.com/apache/nifi/blob/1bd2cf0d09a7111bcecffd0f473aa71c25a69845/nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java),  the script evaluation expects a FlowFile to be returned, in which case it will route the FlowFile to **success**. If a script error occurs, the original FlowFile will be routed to **failure**. 

So, to comply with that statement I think, it is required to add those defaults. 
Please correct me otherwise.

File: nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java (hunk 4 above)
Comment (PuspenduBanerjee) [inline nifi-mock/src/test/java/org/apache/nifi/util/TestMockProcessSession.java]: reverted back.

File: nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java (hunk 6 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: In ExecuteScript, the two relationships success and failure are defined.

In this case, the original code is adding, by default, the two relationships if and only if the processor didn't add its own relationships. With the change you propose, success and failure will be added in any case even if we don't want success/failure, but rather relationships "a", "b", "c". I think that the failing tests should be modified to add the relationships success/failure. At least if I am not missing something here...

Comment (mattyb149) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: +1 to @pvillard31 's comment, that's what I was getting at too :)

Comment (PuspenduBanerjee) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: Okay, I shall revert that into original state and fix test scripts. Thanks for all your guidance and review.

Comment (PuspenduBanerjee) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: Even  in original state of `InvokeScriptedProcesor`, the documentation says 

> SUCCESS and FAILURE are always returned, and if the script
     processor has defined additional relationships, those will be added as well

```
 /**
     * Returns the valid relationships for this processor. SUCCESS and FAILURE are always returned, and if the script
     * processor has defined additional relationships, those will be added as well.
     *
     * @return a Set of Relationships supported by this processor
     */
    @Override
    public Set<Relationship> getRelationships()
```

So, ideally those test cases does not require modification, they are there as per contract stated in the java-doc and this piece is valid.

I strongly believe we should not break that contract as that may create implication for scripted-processor in production. 

Please correct me otherwise.

Comment (PuspenduBanerjee) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: @pvillard31 @mattyb149  any thoughts?

**Comment by PuspenduBanerjee** [comment]
@pvillard31 @mattyb149 @joewitt Any chance to look into this?

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java (hunk 2 above)
Comment (joewitt) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: Instead of rollback what it should do is throw "IllegalArgumentException" as per API documentation found here:
  https://github.com/apache/nifi/blob/master/nifi-api/src/main/java/org/apache/nifi/processor/ProcessSession.java#L382-L383

File: nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java (hunk 6 above)
Comment (joewitt) [inline nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/InvokeScriptedProcessor.java]: @PuspenduBanerjee it looks like the current behavior is correct and as intended as @pvillard31 and @mattyb149 appear to be stating.  If that is true, it looks like the bug here is the documentation.  @mattyb149 please confirm.

**Comment by joewitt** [comment]
I would recommend this PR be closed and instead specific PRs created for each JIRA.

File: nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java (hunk 2 above)
Comment (PuspenduBanerjee) [inline nifi-mock/src/main/java/org/apache/nifi/util/MockProcessSession.java]: I think , @pvillard31  's direction was same too. So, I shall make necessary changes. Can you please check what does Standard Process Session does now a day in such situation?

**Comment by PuspenduBanerjee** [comment]
Closing this PR as per discussion with @joewitt
